source/h2_decoder.c

```diff
@@ -15,48 +15,622 @@
+    (void)succ;
+
+    /* Discard top bit */
+    decoder->frame_in_progress.stream_id = stream_id & s_31_bit_mask;
+
+    const enum aws_h2_frame_type frame_type = decoder->frame_in_progress.type;
+
+    DECODER_LOGF(
+        TRACE,
+        decoder,
+        "Done decoding frame header, beginning to process body of %s frame (stream id=%" PRIu32 "payload len=%" PRIu32
+        " flags=%" PRIu8 ")",
+        aws_h2_frame_type_to_str(frame_type),
+        decoder->frame_in_progress.stream_id,
+        decoder->frame_in_progress.payload_len,
+        decoder->frame_in_progress.flags);
+
+    if (decoder->frame_in_progress.flags & AWS_H2_FRAME_F_PADDED) {
+
+        /* Read padding length if necessary */
+        s_decoder_set_state(decoder, &s_state_padding_len);
+    } else if (decoder->frame_in_progress.flags & AWS_H2_FRAME_T_PRIORITY) {
+
+        /* Read the stream dependency and weight if PRIORITY is set */
+        s_decoder_set_state(decoder, &s_state_priority_block);
+    } else {
+
+        /* Set the state to the appropriate frame's state */
+        s_decoder_set_state(decoder, s_state_frames[frame_type]);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_state_fn_padding_len(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+
+    AWS_FATAL_ASSERT(input->len >= 1);
+
+    const enum aws_h2_frame_type frame_type = decoder->frame_in_progress.type;
+
+    /* Ensure this frame is allowed to be padded */
+    switch (frame_type) {
+        case AWS_H2_FRAME_T_DATA:
+        case AWS_H2_FRAME_T_HEADERS:
+        case AWS_H2_FRAME_T_PUSH_PROMISE:
+            break;
+        default:
+            DECODER_LOGF(ERROR, decoder, "Frame type %s cannot be padded!", aws_h2_frame_type_to_str(frame_type));
+            return aws_raise_error(AWS_ERROR_HTTP_PROTOCOL_ERROR);
+    }
+
+    /* Read the padding length */
+    bool succ = aws_byte_cursor_read_u8(input, &decoder->frame_in_progress.padding_len);
+    AWS_ASSERT(succ);
+    (void)succ;
+
+    /* Adjust payload size */
+    decoder->frame_in_progress.payload_len -= decoder->frame_in_progress.padding_len;
+
+    DECODER_LOGF(
+        TRACE,
+        decoder,
+        "Padding length of %s frame: %" PRIu32,
+        aws_h2_frame_type_to_str(frame_type),
+        decoder->frame_in_progress.padding_len);
+
+    /* Set the state to the appropriate frame's state */
+    s_decoder_set_state(decoder, s_state_frames[frame_type]);
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_state_fn_priority_block(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+
+    AWS_FATAL_ASSERT(input->len >= 5);
+
+    /* Read the top 4 bytes */
+    uint32_t top_bytes = 0;
+    bool succ = aws_byte_cursor_read_be32(input, &top_bytes);
+    AWS_ASSERT(succ);
+    (void)succ;
+
+    /* Write the priority weight */
+    uint8_t weight = 0;
+    succ = aws_byte_cursor_read_u8(input, &weight);
+    AWS_ASSERT(succ);
+    (void)succ;
+
+#if defined(AWS_H2_ENABLE_PRIORITY)
```

Comment:
We should probably assert if someone defines this, until we actually want to support it. It's considered harmful, ATM.

Reply:
I mean, this code won't compile with it, so does that count?